lib/wasi/src/bin_factory/exec.rs

```diff
@@ -195,6 +190,124 @@ pub fn spawn_exec_module(
     Ok(join_handle)
 }
 
+/// Finishes with a module and notifies an errcode
+fn finish_module(env: &WasiEnv, err: ExitCode) {
+    env.blocking_cleanup(Some(err));
+}
+
+/// Calls the module
+fn call_module(
+    ctx: WasiFunctionEnv,
+    mut store: Store,
+    module: Module,
+    start: Function,
+    rewind_state: Option<(RewindState, Result<(), Errno>)>,
+) {
+    let env = ctx.data(&store);
+    let pid = env.pid();
+    let thread = env.thread.clone();
+    let tasks = env.tasks().clone();
+    thread.set_status_running();
+
+    // If we need to rewind then do so
+    if let Some((mut rewind_state, trigger_res)) = rewind_state {
```

Comment:
This should only be done if async threads are enabled.

Reply:
refactored, needs new review